File: src/volumeAnalysis/islandVolume/steps/voxelization/rle.ts
import { ComponentInfo } from './types';

// RLE Row: [start, length, start, length, ...]
// Represents a single row of binary data.
export type RleRow = Int32Array;

// RLE Mask: Array of rows
export type RleMask = {
    rows: RleRow[];
    width: number;
    height: number;
};

// Labeled RLE Row: [start, length, id, start, length, id, ...]
export type RleLabelRow = Int32Array;

export type RleLabels = {
    rows: RleLabelRow[];
    width: number;
    height: number;
};

/**
 * Encodes a binary grid into RLE format.
 */
export function rleEncode(data: Uint8Array, width: number, height: number): RleMask {
    const rows: RleRow[] = new Array(height);

    for (let y = 0; y < height; y++) {
        const rowSpans: number[] = [];
        let runStart = -1;
        const rowOffset = y * width;

        for (let x = 0; x < width; x++) {
            if (data[rowOffset + x]) {
                if (runStart === -1) runStart = x;
            } else {
                if (runStart !== -1) {
                    rowSpans.push(runStart, x - runStart);
                    runStart = -1;
                }
            }
        }
        if (runStart !== -1) {
            rowSpans.push(runStart, width - runStart);
        }
        rows[y] = new Int32Array(rowSpans);
    }

    return { rows, width, height };
}

/**
 * Encodes a dense ID grid into RLE Label format.
 * Preserves the ID in each run: [start, length, id]
 */
export function rleEncodeLabels(data: Int32Array | Uint8Array, width: number, height: number): RleLabels {
    const rows: RleLabelRow[] = new Array(height);

    for (let y = 0; y < height; y++) {
        const rowSpans: number[] = [];
        let runStart = -1;
        let currentId = 0;
        const rowOffset = y * width;

        for (let x = 0; x < width; x++) {
            const id = data[rowOffset + x];

            if (id !== currentId) {
                // Change detected
                if (currentId !== 0 && runStart !== -1) {
                    // Close previous run
                    rowSpans.push(runStart, x - runStart, currentId);
                }

                // Start new run (if non-zero)
                if (id !== 0) {
                    runStart = x;
                } else {
                    runStart = -1;
                }
                currentId = id;
            }
        }

        // Close final run
        if (currentId !== 0 && runStart !== -1) {
            rowSpans.push(runStart, width - runStart, currentId);
        }

        rows[y] = new Int32Array(rowSpans);
    }

    return { rows, width, height };
}

/**
 * Decodes RLE format back to binary grid.
 */
export function rleDecode(mask: RleMask): Uint8Array {
    const { rows, width, height } = mask;
    const data = new Uint8Array(width * height);

    for (let y = 0; y < height; y++) {
        const row = rows[y];
        const rowOffset = y * width;
        for (let i = 0; i < row.length; i += 2) {
            const start = row[i];
            const len = row[i + 1];
            for (let j = 0; j < len; j++) {
                data[rowOffset + start + j] = 1;
            }
        }
    }

    return data;
}

/**
 * Computes Intersection of A and Dilated(B).
 * Result = A AND Dilate(B, buffer)
 * Used to find "Supported" regions.
 */
export function rleIntersectDilated(a: RleMask, b: RleMask, buffer: number): RleMask {
    const { width, height } = a;
    const resultRows: RleRow[] = new Array(height);

    // Optimization: Pre-calculate dilated bounds for B rows?
    // Or just iterate.

    for (let y = 0; y < height; y++) {
        const aRow = a.rows[y];
        if (aRow.length === 0) {
            resultRows[y] = new Int32Array(0);
            continue;
        }

        // Gather relevant rows from B (y +/- buffer)
        // We treat B as if every run is expanded by 'buffer' in X and Y.
        // So a run in B at 'by' covers Y range [by - buffer, by + buffer].
        // Conversely, for current Y, we look at B rows in [y - buffer, y + buffer].

        const relevantBRows: RleRow[] = [];
        const startY = Math.max(0, y - buffer);
        const endY = Math.min(height - 1, y + buffer);

        for (let by = startY; by <= endY; by++) {
            if (b.rows[by].length > 0) {
                relevantBRows.push(b.rows[by]);
            }
        }

        if (relevantBRows.length === 0) {
            resultRows[y] = new Int32Array(0);
            continue;
        }

        // Intersect 'aRow' with the union of 'relevantBRows' (dilated horizontally)
        // We can merge relevantBRows into a single sorted list of intervals, merging overlaps,
        // and expanding each by 'buffer'.

        const bIntervals: { start: number, end: number }[] = [];
        for (const bRow of relevantBRows) {
            for (let i = 0; i < bRow.length; i += 2) {
                // Dilate horizontally
                bIntervals.push({
                    start: Math.max(0, bRow[i] - buffer),
                    end: Math.min(width, bRow[i] + bRow[i + 1] + buffer)
                });
            }
        }

        // Sort and merge B intervals
        bIntervals.sort((p, q) => p.start - q.start);

        const mergedB: { start: number, end: number }[] = [];
        if (bIntervals.length > 0) {
            let curr = bIntervals[0];
            for (let i = 1; i < bIntervals.length; i++) {
                const next = bIntervals[i];
                if (next.start <= curr.end) {
                    curr.end = Math.max(curr.end, next.end);
                } else {
                    mergedB.push(curr);
                    curr = next;
                }
            }
            mergedB.push(curr);
        }

        // Now intersect aRow with mergedB
        const resSpans: number[] = [];
        let bIdx = 0;

        for (let i = 0; i < aRow.length; i += 2) {
            const aStart = aRow[i];
            const aEnd = aStart + aRow[i + 1];

            // Advance bIdx
            while (bIdx < mergedB.length && mergedB[bIdx].end <= aStart) {
                bIdx++;
            }

            let tempBIdx = bIdx;
            while (tempBIdx < mergedB.length && mergedB[tempBIdx].start < aEnd) {
                const bInt = mergedB[tempBIdx];
                const start = Math.max(aStart, bInt.start);
                const end = Math.min(aEnd, bInt.end);

                if (start < end) {
                    // Append run, merging if adjacent
                    if (resSpans.length > 0 && resSpans[resSpans.length - 2] + resSpans[resSpans.length - 1] === start) {
                        resSpans[resSpans.length - 1] += (end - start);
                    } else {
                        resSpans.push(start, end - start);
                    }
                }
                tempBIdx++;
            }
        }

        resultRows[y] = new Int32Array(resSpans);
    }

    return { rows: resultRows, width, height };
}

/**
 * Computes A MINUS B.
 * Result = A AND NOT B
 */
export function rleSubtract(a: RleMask, b: RleMask): RleMask {
    const { width, height } = a;
    const resultRows: RleRow[] = new Array(height);

    for (let y = 0; y < height; y++) {
        const aRow = a.rows[y];
        const bRow = b.rows[y];

        if (aRow.length === 0) {
            resultRows[y] = new Int32Array(0);
            continue;
        }
        if (bRow.length === 0) {
            resultRows[y] = aRow; // Copy? Or ref is fine if immutable
            continue;
        }

        const resSpans: number[] = [];
        let bIdx = 0;

        for (let i = 0; i < aRow.length; i += 2) {
            let currentStart = aRow[i];
            const currentEnd = currentStart + aRow[i + 1];

            // Skip B runs that end before current A run
            while (bIdx < bRow.length && bRow[bIdx] + bRow[bIdx + 1] <= currentStart) {
                bIdx += 2;
            }

            // Check overlaps
            let tempBIdx = bIdx;
            while (tempBIdx < bRow.length && bRow[tempBIdx] < currentEnd) {
                const bStart = bRow[tempBIdx];
                const bEnd = bStart + bRow[tempBIdx + 1];

                if (bStart > currentStart) {
                    // Add gap before B
                    resSpans.push(currentStart, bStart - currentStart);
                }

                currentStart = Math.max(currentStart, bEnd);
                tempBIdx += 2;
            }

            if (currentStart < currentEnd) {
                resSpans.push(currentStart, currentEnd - currentStart);
            }
        }

        resultRows[y] = new Int32Array(resSpans);
    }

    return { rows: resultRows, width, height };
}

/**
 * Connected Components Labeling on RLE Mask.
 * Returns labeled RLE and component info.
 */
export function rleLabelComponents(mask: RleMask, connectivity: 4 | 8 = 4): { labels: RleLabels, components: ComponentInfo[] } {
    const { rows, width, height } = mask;
    const labelRows: RleLabelRow[] = new Array(height);

    // Union-Find structure
    const parent: number[] = [0]; // 1-based IDs
    const area: number[] = [0];
    const sumX: number[] = [0];
    const sumY: number[] = [0];
    let nextId = 1;

    function find(i: number): number {
        if (parent[i] === i) return i;
        parent[i] = find(parent[i]);
        return parent[i];
    }

    function union(i: number, j: number) {
        const rootI = find(i);
        const rootJ = find(j);
        if (rootI !== rootJ) {
            parent[rootJ] = rootI;
            area[rootI] += area[rootJ];
            sumX[rootI] += sumX[rootJ];
            sumY[rootI] += sumY[rootJ];
            area[rootJ] = 0; // Optional, just to be clean
            sumX[rootJ] = 0;
            sumY[rootJ] = 0;
        }
    }

    function newLabel(initialArea: number, initialSumX: number, initialSumY: number): number {
        const id = nextId++;
        parent[id] = id;
        area[id] = initialArea;
        sumX[id] = initialSumX;
        sumY[id] = initialSumY;
        return id;
    }

    // First pass: assign temporary labels and merge connected runs
    // We store temp labels directly in labelRows

    for (let y = 0; y < height; y++) {
        const row = rows[y];
        const prevRow = y > 0 ? labelRows[y - 1] : null;
        const currentRowLabels: number[] = [];
        let pIdx = 0; // Pointer for prevRow optimization

        for (let i = 0; i < row.length; i += 2) {
            const start = row[i];
            const len = row[i + 1];
            const end = start + len;

            // Calculate sums for this run
            const runSumX = len * (start + (end - 1)) / 2;
            const runSumY = len * y;

            // Create new label for this run
            const myId = newLabel(len, runSumX, runSumY);
            currentRowLabels.push(start, len, myId);

            // Check connectivity with previous row
            if (prevRow) {
                const expand = connectivity === 8 ? 1 : 0;
                const searchStart = start - expand;
                const searchEnd = end + expand;

                // OPTIMIZATION: Use sliding pointer pIdx to avoid O(N^2)
                // Advance pIdx to skip runs that end before searchStart
                while (pIdx < prevRow.length) {
                    const pStart = prevRow[pIdx];
                    const pLen = prevRow[pIdx + 1];
                    const pEnd = pStart + pLen;
                    if (pEnd <= searchStart) {
                        pIdx += 3; // label rows are [start, len, id]
                    } else {
                        break;
                    }
                }

                // Check overlaps starting from pIdx
                let tempIdx = pIdx;
                while (tempIdx < prevRow.length) {
                    const pStart = prevRow[tempIdx];
                    // const pLen = prevRow[tempIdx + 1]; // Unused
                    const pId = prevRow[tempIdx + 2];

                    if (pStart >= searchEnd) break; // Passed the window

                    // Overlap confirmed by loop bounds (since we skipped non-overlapping left ones)
                    union(myId, pId);

                    tempIdx += 3;
                }
            }
        }
        labelRows[y] = new Int32Array(currentRowLabels);
    }

    // Second pass: resolve labels and build component list
    const finalComponents: ComponentInfo[] = [];
    const idMap = new Map<number, number>(); // Root ID -> Final Component ID

    // Re-map IDs to be sequential 1..N
    let finalNextId = 1;

    for (let y = 0; y < height; y++) {
        const row = labelRows[y];
        for (let i = 0; i < row.length; i += 3) {
            const oldId = row[i + 2];
            const rootId = find(oldId);

            let finalId = idMap.get(rootId);
            if (finalId === undefined) {
                finalId = finalNextId++;
                idMap.set(rootId, finalId);
                finalComponents.push({
                    id: finalId,
                    label: finalId,
                    area_px: area[rootId],
                    size: area[rootId],
                    centroidSumX: sumX[rootId],
                    centroidSumY: sumY[rootId]
                });
            }

            row[i + 2] = finalId;
        }
    }

    return { labels: { rows: labelRows, width, height }, components: finalComponents };
}

/**
 * Decodes RLE Labels back to dense integer grid.
 */
export function rleDecodeLabels(labels: RleLabels): Int32Array {
    const { rows, width, height } = labels;
    const data = new Int32Array(width * height);

    for (let y = 0; y < height; y++) {
        const row = rows[y];
        const rowOffset = y * width;
        for (let i = 0; i < row.length; i += 3) {
            const start = row[i];
            const len = row[i + 1];
            const id = row[i + 2];
            for (let j = 0; j < len; j++) {
                data[rowOffset + start + j] = id;
            }
        }
    }

    return data;
}

File: rust/dragonfruit-islands/src/rle.rs
//! RLE Domain Service — stateless mask algebra.
//!
//! Ported from `IslandScan/rle.ts`. All operations are pure functions on
//! RLE Value Objects.

use crate::model::*;

// ---------------------------------------------------------------------------
// Encode / Decode
// ---------------------------------------------------------------------------

/// Encode a dense binary grid (0/non-0) into an `RleMask`.
pub fn rle_encode(data: &[u8], width: i32, height: i32) -> RleMask {
    let w = width as usize;
    let mut rows = Vec::with_capacity(height as usize);

    for y in 0..height as usize {
        let mut spans = Vec::new();
        let row_off = y * w;
        let mut run_start: Option<i32> = None;

        for x in 0..w {
            if data[row_off + x] != 0 {
                if run_start.is_none() {
                    run_start = Some(x as i32);
                }
            } else if let Some(s) = run_start {
                spans.push(RleRun {
                    start: s,
                    length: x as i32 - s,
                });
                run_start = None;
            }
        }
        if let Some(s) = run_start {
            spans.push(RleRun {
                start: s,
                length: width - s,
            });
        }
        rows.push(spans);
    }

    RleMask {
        rows,
        width,
        height,
    }
}

/// Decode an `RleMask` back to a dense binary grid (`0` / `1`).
pub fn rle_decode(mask: &RleMask) -> Vec<u8> {
    let w = mask.width as usize;
    let mut data = vec![0u8; w * mask.height as usize];

    for (y, row) in mask.rows.iter().enumerate() {
        let row_off = y * w;
        for run in row {
            for j in 0..run.length {
                data[row_off + (run.start + j) as usize] = 1;
            }
        }
    }
    data
}

/// Encode a dense label grid (i32, 0 = background) into `RleLabels`.
pub fn rle_encode_labels(data: &[i32], width: i32, height: i32) -> RleLabels {
    let w = width as usize;
    let mut rows = Vec::with_capacity(height as usize);

    for y in 0..height as usize {
        let mut spans = Vec::new();
        let row_off = y * w;
        let mut run_start: i32 = -1;
        let mut current_id: i32 = 0;

        for x in 0..w {
            let id = data[row_off + x];
            if id != current_id {
                if current_id != 0 && run_start >= 0 {
                    spans.push(RleLabelRun {
                        start: run_start,
                        length: x as i32 - run_start,
                        id: current_id,
                    });
                }
                if id != 0 {
                    run_start = x as i32;
                } else {
                    run_start = -1;
                }
                current_id = id;
            }
        }
        if current_id != 0 && run_start >= 0 {
            spans.push(RleLabelRun {
                start: run_start,
                length: width - run_start,
                id: current_id,
            });
        }
        rows.push(spans);
    }

    RleLabels {
        rows,
        width,
        height,
    }
}

/// Decode `RleLabels` to a dense i32 grid.
pub fn rle_decode_labels(labels: &RleLabels) -> Vec<i32> {
    let w = labels.width as usize;
    let mut data = vec![0i32; w * labels.height as usize];

    for (y, row) in labels.rows.iter().enumerate() {
        let row_off = y * w;
        for run in row {
            for j in 0..run.length {
                data[row_off + (run.start + j) as usize] = run.id;
            }
        }
    }
    data
}

// ---------------------------------------------------------------------------
// Intersect Dilated: Result = A AND Dilate(B, buffer)
// ---------------------------------------------------------------------------

/// Compute `A AND Dilate(B, buffer)`.
///
/// Used to find "supported" regions: current layer pixels that overlap with
/// the dilated previous layer.
pub fn rle_intersect_dilated(a: &RleMask, b: &RleMask, buffer: i32) -> RleMask {
    let width = a.width;
    let height = a.height;
    let mut result_rows = Vec::with_capacity(height as usize);

    for y in 0..height {
        let a_row = &a.rows[y as usize];
        if a_row.is_empty() {
            result_rows.push(Vec::new());
            continue;
        }

        // Gather relevant B rows within [y - buffer, y + buffer]
        let start_y = (y - buffer).max(0);
        let end_y = (y + buffer).min(height - 1);

        // Collect all dilated B intervals
        let mut b_intervals: Vec<(i32, i32)> = Vec::new();
        for by in start_y..=end_y {
            for run in &b.rows[by as usize] {
                let s = (run.start - buffer).max(0);
                let e = (run.start + run.length + buffer).min(width);
                b_intervals.push((s, e));
            }
        }

        if b_intervals.is_empty() {
            result_rows.push(Vec::new());
            continue;
        }

        // Sort and merge B intervals
        b_intervals.sort_unstable_by_key(|&(s, _)| s);
        let merged_b = merge_intervals(&b_intervals);

        // Intersect A row with merged B intervals
        let mut res = Vec::new();
        let mut bi = 0;

        for a_run in a_row {
            let a_start = a_run.start;
            let a_end = a_start + a_run.length;

            // Advance bi past intervals that end before a_start
            while bi < merged_b.len() && merged_b[bi].1 <= a_start {
                bi += 1;
            }

            let mut ti = bi;
            while ti < merged_b.len() && merged_b[ti].0 < a_end {
                let start = a_start.max(merged_b[ti].0);
                let end = a_end.min(merged_b[ti].1);
                if start < end {
                    // Merge with previous run if adjacent
                    if let Some(last) = res.last_mut() {
                        let last_run: &mut RleRun = last;
                        if last_run.start + last_run.length == start {
                            last_run.length += end - start;
                        } else {
                            res.push(RleRun {
                                start,
                                length: end - start,
                            });
                        }
                    } else {
                        res.push(RleRun {
                            start,
                            length: end - start,
                        });
                    }
                }
                ti += 1;
            }
        }

        result_rows.push(res);
    }

    RleMask {
        rows: result_rows,
        width,
        height,
    }
}

// ---------------------------------------------------------------------------
// Subtract: Result = A AND NOT B
// ---------------------------------------------------------------------------

/// Compute `A MINUS B` (= `A AND NOT B`).
pub fn rle_subtract(a: &RleMask, b: &RleMask) -> RleMask {
    let width = a.width;
    let height = a.height;
    let mut result_rows = Vec::with_capacity(height as usize);

    for y in 0..height as usize {
        let a_row = &a.rows[y];
        let b_row = &b.rows[y];

        if a_row.is_empty() {
            result_rows.push(Vec::new());
            continue;
        }
        if b_row.is_empty() {
            result_rows.push(a_row.clone());
            continue;
        }

        let mut res = Vec::new();
        let mut bi = 0;

        for a_run in a_row {
            let mut cur_start = a_run.start;
            let cur_end = a_run.start + a_run.length;

            // Skip B runs that end before current A run
            while bi < b_row.len() && b_row[bi].start + b_row[bi].length <= cur_start {
                bi += 1;
            }

            let mut tbi = bi;
            while tbi < b_row.len() && b_row[tbi].start < cur_end {
                let b_start = b_row[tbi].start;
                let b_end = b_start + b_row[tbi].length;

                if b_start > cur_start {
                    res.push(RleRun {
                        start: cur_start,
                        length: b_start - cur_start,
                    });
                }
                cur_start = cur_start.max(b_end);
                tbi += 1;
            }

            if cur_start < cur_end {
                res.push(RleRun {
                    start: cur_start,
                    length: cur_end - cur_start,
                });
            }
        }

        result_rows.push(res);
    }

    RleMask {
        rows: result_rows,
        width,
        height,
    }
}

// ---------------------------------------------------------------------------
// Connected Components Labeling
// ---------------------------------------------------------------------------

/// Connected component labeling on an `RleMask`.
///
/// Two-pass algorithm using union-find over RLE runs.
/// Returns labeled RLE and per-component metadata.
pub fn rle_label_components(
    mask: &RleMask,
    connectivity: Connectivity,
) -> (RleLabels, Vec<ComponentInfo>) {
    let height = mask.height as usize;
    let expand = if connectivity == Connectivity::Eight {
        1
    } else {
        0
    };

    // Union-find arrays (1-based)
    let mut parent: Vec<usize> = vec![0]; // index 0 unused
    let mut area: Vec<i64> = vec![0];
    let mut sum_x: Vec<f64> = vec![0.0];
    let mut sum_y: Vec<f64> = vec![0.0];
    let mut next_id: usize = 1;

    // Path-compressed find
    fn find(parent: &mut [usize], mut i: usize) -> usize {
        while parent[i] != i {
            parent[i] = parent[parent[i]];
            i = parent[i];
        }
        i
    }

    let union = |parent: &mut Vec<usize>,
                 area: &mut Vec<i64>,
                 sum_x: &mut Vec<f64>,
                 sum_y: &mut Vec<f64>,
                 i: usize,
                 j: usize| {
        let ri = find(parent, i);
        let rj = find(parent, j);
        if ri != rj {
            parent[rj] = ri;
            area[ri] += area[rj];
            sum_x[ri] += sum_x[rj];
            sum_y[ri] += sum_y[rj];
            area[rj] = 0;
            sum_x[rj] = 0.0;
            sum_y[rj] = 0.0;
        }
    };

    // First pass: assign temp labels, union connected runs
    let mut label_rows: Vec<Vec<RleLabelRun>> = Vec::with_capacity(height);

    for y in 0..height {
        let row = &mask.rows[y];
        let mut current_row_labels = Vec::new();

        for run in row {
            let start = run.start;
            let len = run.length;
            let end = start + len;

            // Centroid sums for this run
            let run_sum_x = len as f64 * (start as f64 + (end - 1) as f64) / 2.0;
            let run_sum_y = len as f64 * y as f64;

            // Allocate new label
            let my_id = next_id;
            next_id += 1;
            parent.push(my_id);
            area.push(len as i64);
            sum_x.push(run_sum_x);
            sum_y.push(run_sum_y);

            current_row_labels.push(RleLabelRun {
                start,
                length: len,
                id: my_id as i32,
            });

            // Check connectivity with previous row
            if y > 0 {
                let prev_row = &label_rows[y - 1];
                let search_start = start - expand;
                let search_end = end + expand;

                for prev_run in prev_row {
                    let p_start = prev_run.start;
                    let p_end = p_start + prev_run.length;

                    if p_start >= search_end {
                        break;
                    }
                    if search_start.max(p_start) < search_end.min(p_end) {
                        union(
                            &mut parent,
                            &mut area,
                            &mut sum_x,
                            &mut sum_y,
                            my_id,
                            prev_run.id as usize,
                        );
                    }
                }
            }
        }

        label_rows.push(current_row_labels);
    }

    // Second pass: resolve labels to sequential IDs, build components
    let mut id_map: HashMap<usize, i32> = HashMap::new();
    let mut components = Vec::new();
    let mut final_next_id: i32 = 1;

    use std::collections::HashMap;

    for y in 0..height {
        for run in &mut label_rows[y] {
            let old_id = run.id as usize;
            let root = find(&mut parent, old_id);

            let final_id = *id_map.entry(root).or_insert_with(|| {
                let fid = final_next_id;
                final_next_id += 1;
                components.push(ComponentInfo {
                    id: fid,
                    label: fid,
                    area_px: area[root] as i32,
                    size: area[root] as i32,
                    centroid_sum_x: sum_x[root],
                    centroid_sum_y: sum_y[root],
                });
                fid
            });

            run.id = final_id;
        }
    }

    let labels = RleLabels {
        rows: label_rows,
        width: mask.width,
        height: mask.height,
    };

    (labels, components)
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn merge_intervals(intervals: &[(i32, i32)]) -> Vec<(i32, i32)> {
    if intervals.is_empty() {
        return Vec::new();
    }
    let mut merged = Vec::with_capacity(intervals.len());
    let mut cur = intervals[0];
    for &(s, e) in &intervals[1..] {
        if s <= cur.1 {
            cur.1 = cur.1.max(e);
        } else {
            merged.push(cur);
            cur = (s, e);
        }
    }
    merged.push(cur);
    merged
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use crate::model::*;
    use crate::rle::*;

    #[test]
    fn encode_decode_roundtrip() {
        let w = 8;
        let h = 4;
        #[rustfmt::skip]
        let data: Vec<u8> = vec![
            0,0,1,1,1,0,0,0,
            0,1,1,0,0,1,1,0,
            0,0,0,0,0,0,0,0,
            1,1,1,1,1,1,1,1,
        ];
        let mask = rle_encode(&data, w, h);
        assert_eq!(mask.rows[0].len(), 1); // one run: [2,3]
        assert_eq!(mask.rows[1].len(), 2); // two runs
        assert_eq!(mask.rows[2].len(), 0); // empty
        assert_eq!(mask.rows[3].len(), 1); // one full-width run

        let decoded = rle_decode(&mask);
        assert_eq!(decoded, data);
    }

    #[test]
    fn encode_decode_labels_roundtrip() {
        let w = 6;
        let h = 2;
        let data: Vec<i32> = vec![0, 1, 1, 0, 2, 2, 3, 3, 3, 0, 0, 2];
        let labels = rle_encode_labels(&data, w, h);
        let decoded = rle_decode_labels(&labels);
        assert_eq!(decoded, data);
    }

    #[test]
    fn subtract_basic() {
        let w = 10;
        let h = 1;
        let a = RleMask {
            rows: vec![vec![RleRun {
                start: 2,
                length: 6,
            }]],
            width: w,
            height: h,
        };
        let b = RleMask {
            rows: vec![vec![RleRun {
                start: 4,
                length: 2,
            }]],
            width: w,
            height: h,
        };
        let result = rle_subtract(&a, &b);
        // Expected: [2,2] and [6,2]
        assert_eq!(result.rows[0].len(), 2);
        assert_eq!(result.rows[0][0], RleRun { start: 2, length: 2 });
        assert_eq!(result.rows[0][1], RleRun { start: 6, length: 2 });
    }

    #[test]
    fn intersect_dilated_basic() {
        let w = 10;
        let h = 3;
        let a = RleMask {
            rows: vec![
                vec![RleRun {
                    start: 0,
                    length: 10,
                }],
                vec![RleRun {
                    start: 0,
                    length: 10,
                }],
                vec![RleRun {
                    start: 0,
                    length: 10,
                }],
            ],
            width: w,
            height: h,
        };
        // B has a single pixel at (5, 1)
        let b = RleMask {
            rows: vec![
                vec![],
                vec![RleRun {
                    start: 5,
                    length: 1,
                }],
                vec![],
            ],
            width: w,
            height: h,
        };
        // buffer=1: dilated B covers [4,7) across rows 0-2
        let result = rle_intersect_dilated(&a, &b, 1);
        // All three rows should have a run [4, 3]
        for row in &result.rows {
            assert_eq!(row.len(), 1);
            assert_eq!(row[0], RleRun { start: 4, length: 3 });
        }
    }

    #[test]
    fn label_components_4_connectivity() {
        let w = 5;
        let h = 3;
        #[rustfmt::skip]
        let data: Vec<u8> = vec![
            1,1,0,1,1,
            0,0,0,0,0,
            1,1,0,0,1,
        ];
        let mask = rle_encode(&data, w, h);
        let (labels, comps) = rle_label_components(&mask, Connectivity::Four);

        // Should find 4 disconnected components (4-conn):
        // (0-1, row 0), (3-4, row 0), (0-1, row 2), (4, row 2)
        assert_eq!(comps.len(), 4);
        assert_eq!(labels.rows.len(), 3);

        // Total area should match
        let total: i32 = comps.iter().map(|c| c.area_px).sum();
        assert_eq!(total, 7);
    }

    #[test]
    fn label_components_8_connectivity() {
        let w = 5;
        let h = 3;
        #[rustfmt::skip]
        let data: Vec<u8> = vec![
            1,0,0,0,0,
            0,1,0,0,0,
            0,0,1,0,0,
        ];
        let mask = rle_encode(&data, w, h);
        let (_labels, comps) = rle_label_components(&mask, Connectivity::Eight);

        // With 8-connectivity, these diagonal pixels form 1 component
        assert_eq!(comps.len(), 1);
        assert_eq!(comps[0].area_px, 3);
    }

    #[test]
    fn pixel_count() {
        let mask = RleMask {
            rows: vec![
                vec![RleRun { start: 0, length: 5 }],
                vec![],
                vec![
                    RleRun { start: 2, length: 3 },
                    RleRun { start: 7, length: 2 },
                ],
            ],
            width: 10,
            height: 3,
        };
        assert_eq!(mask.pixel_count(), 10);
    }
}
